src/clj/bluegenes_tool_store/tools.clj

```diff
@@ -1,78 +1,206 @@
+      install    (partial pkg/install-package   tools-config tools-path)
+      uninstall  (partial pkg/uninstall-package tools-config tools-path)]
+  (defn package-operation
+    "Perform a package install or uninstall `operation` on `package+` which can
+    be a single package name or a collection of multiple. Returns an error
+    response if a different package-operation is already in progress.
+    On success it will return a response with the new tool list."
+    [operation package+]
+    (if (.tryLock lock)
+      (try
+        (case operation
+          :install (if (coll? package+)
+                     (doseq [package-name package+]
+                       (install package-name))
+                     (install package+))
+          :uninstall (if (coll? package+)
+                       (doseq [package-name package+]
+                         (uninstall package-name))
+                       (uninstall package+)))
+        (tools-list-res)
+        (catch Exception e
+          (error e)
+          (response/bad-gateway {:error (.getMessage e)}))
+        (finally
+          (.unlock lock)))
+      (response/service-unavailable {:error "A tool operation is already in progress. Please try again later."}))))
+
+(defn verify-package-params
+  "Helper to wrap API functions to check for validity of package parameters."
+  [f]
+  (fn [{:keys [params] :as req}]
+    (if (or (contains? params :package)
+            (contains? params :packages))
+      (f req)
+      (response/bad-request {:error "Please include your package name(s) in the package or packages parameter."}))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;
+;; Tool API Endpoints ;;
+;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-all-tools
+  "Return list of tools as a REST response to our GET."
+  []
+  (tools-list-res))
+
+(defn get-tools-path
+  "Respond with tool path."
+  []
+  (response/ok {:path (.getAbsolutePath tools-path)}))
+
+(def install-tool
+  "Takes a request to install one or more tools."
+  (verify-package-params
+    (fn [{{:keys [package packages]} :params}]
+      (package-operation :install (or package packages)))))
+
+(def uninstall-tool
+  "Takes a request to uninstall a tool."
+  (verify-package-params
+    (fn [{{:keys [package packages]} :params}]
+      (package-operation :uninstall (or package packages)))))
+
+(def update-tools
+  "Takes a request to update a list of tools."
+  (verify-package-params
+    (fn [{{:keys [package packages]} :params}]
+      (package-operation :install (or package packages)))))
+
+;;;;;;;;;;;;;;;
+;; CLI Usage ;;
+;;;;;;;;;;;;;;;
+
+(defn -main
+  "This main function is for managing tools from the command line via the
+  leiningen `lein tools` alias."
+  [cmd & args]
+  (timbre/set-level! :info)
+  (case cmd
+    "init"      (force-initialise-tools)
+    "install"   (package-operation :install args)
+    "uninstall" (package-operation :uninstall args)
+    (println "Install BlueGenes tools from the command line.
+
+USAGE: lein tools [ARGS...]
```

Comment:
This is gorgeous! It should be in the readme, as a link to this comment if nothing else.

Reply:
Just a note: the `lein tools` commands won't give you good feedback as the error handling was made with the tool store in mind (ie. installing a nonexistant package will give you a "network error" and installing a package succesfully will just silently complete). I thought this was alright since the commands would just be used temporarily by us to manage *http://bluegenes.apps.intermine.org/* until flymine gets updated.
Do you still think we should mention it in the readme?